src/compiler/checker.ts

```diff
@@ -17651,12 +17652,51 @@ namespace ts {
             }
         }
 
+        function getOptionalTypeRecursive(type: Type) {
+            if ((type as ResolvedType).properties) {
+                const rType = type as ResolvedType;
+                const propCount = rType.properties.length;
+
+                if (rType.flags & TypeFlags.Object) {
+                  for (let i = 0; i < propCount; i++) {

```

Comment:
`for (const prop in getPropertiesOfType(rType))`

This might also collapse the object vs union/intersection code paths into one.

Reply:
I will look into that.